<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Integrative genomics viewer</title>

<script src="lab_visualization_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="lab_visualization_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="lab_visualization_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="lab_visualization_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="lab_visualization_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="lab_visualization_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="lab_visualization_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="lab_visualization_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="lab_visualization_files/navigation-1.1/tabsets.js"></script>
<link href="lab_visualization_files/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="lab_visualization_files/highlightjs-9.12.0/highlight.js"></script>
<link href="lab_visualization_files/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="lab_visualization_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="lab_visualization_files/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
<link href="lab_visualization_files/vembedr-0.1.5/css/vembedr.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:20px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Integrative genomics viewer</h1>
<h4 class="date">21 September, 2022</h4>

</div>


<!-- rmd lab header -->
<p><br></p>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<ul>
<li><p>Data visualization is an essential component of genomic data analysis. However, the size and diversity of the data sets produced by today’s sequencing and array-based profiling methods present major challenges to visualization tools.</p></li>
<li><p>The Integrative Genomics Viewer (IGV) is a high-performance viewer that efficiently handles large heterogeneous data sets, while providing a smooth and intuitive user experience at all levels of genome resolution.</p></li>
<li><p>A key characteristic of IGV is its focus on the integrative nature of genomic studies, with support for both array-based and next-generation sequencing data, and the integration of clinical and phenotypic data.</p></li>
<li><p>Although IGV is often used to view genomic data from public sources, its primary emphasis is to support researchers who wish to visualize and explore their own data sets or those from colleagues.</p></li>
<li><p>IGV supports flexible loading of local and remote data sets, and is optimized to provide high-performance data visualization and exploration on standard desktop systems.</p></li>
<li><p>IGV is freely available for <a href="http://www.broadinstitute.org/igv">download</a> under a GNU LGPL open-source license.</p></li>
</ul>
</div>
<div id="objectives" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Objectives</h1>
<ul>
<li><a href="#install-igv">Install IGV</a></li>
<li><a href="#igv-videos">IGV videos</a></li>
<li><a href="#batch-commands">Batch commands</a></li>
<li><a href="#reference-sheet">Reference sheet</a></li>
</ul>
</div>
<div id="tutorial" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Tutorial</h1>
<p>In this tutorial we are going to learn how to use IGV to visualize genomic data. The first thing to do is create a directory to store all the tutorial data. It is good practice to create a new directory for each project you work on, this ensures files do not get mixed up and all the results are self-contained.</p>
<p>Create a ‘tutorial’ directory to store output files:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>mkdir tutorial</code></pre>
</div>
<p>Download the tutorial and exercise data:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>curl https://raw.githubusercontent.com/zifornd/workshop-template-rmd-ga/master/data/visualization/data.tar.gz --output tutorial/data.tar.gz</code></pre>
<pre><code>##   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
##                                  Dload  Upload   Total   Spent    Left  Speed
## 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  6 13.8M    6  927k    0     0   751k      0  0:00:18  0:00:01  0:00:17  755k
 20 13.8M   20 2927k    0     0  1310k      0  0:00:10  0:00:02  0:00:08 1313k
 34 13.8M   34 4943k    0     0  1529k      0  0:00:09  0:00:03  0:00:06 1531k
 48 13.8M   48 6927k    0     0  1636k      0  0:00:08  0:00:04  0:00:04 1638k
 62 13.8M   62 8896k    0     0  1699k      0  0:00:08  0:00:05  0:00:03 1780k
 77 13.8M   77 10.6M    0     0  1750k      0  0:00:08  0:00:06  0:00:02 1996k
 91 13.8M   91 12.6M    0     0  1784k      0  0:00:07  0:00:07 --:--:-- 1996k
100 13.8M  100 13.8M    0     0  1804k      0  0:00:07  0:00:07 --:--:-- 1997k</code></pre>
</div>
<p>Extract the archive file into the tutorial directory:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>tar xf tutorial/data.tar.gz --directory=tutorial</code></pre>
</div>
<div id="install-igv" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Install IGV</h2>
<p>The software we are going to use in this tutorial can be installed using the conda package manager. Please refer to the previous conda workshop for details on installing software and creating conda environments.</p>
<p>Create a new environment with IGV installed:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>conda create --yes --name igvtools igv</code></pre>
<pre><code>## Collecting package metadata (current_repodata.json): ...working... done
## Solving environment: ...working... done
## 
## ## Package Plan ##
## 
##   environment location: /opt/miniconda3/envs/igvtools
## 
##   added / updated specs:
##     - igv
## 
## 
## The following NEW packages will be INSTALLED:
## 
##   igv                bioconda/noarch::igv-2.13.2-hdfd78af_0
##   libcxx             conda-forge/osx-64::libcxx-14.0.6-hccf4f1f_0
##   libzlib            conda-forge/osx-64::libzlib-1.2.12-hfd90126_3
##   openjdk            conda-forge/osx-64::openjdk-17.0.3-hbc0c0cd_2
## 
## 
## Preparing transaction: ...working... done
## Verifying transaction: ...working... done
## Executing transaction: ...working... done
## #
## # To activate this environment, use
## #
## #     $ conda activate igvtools
## #
## # To deactivate an active environment, use
## #
## #     $ conda deactivate
## 
## Retrieving notices: ...working... done</code></pre>
</div>
<!-- On macOS, the IGV launcher requires the coreutils version of readlink. After installing the conda environment, manually change the path to the readlink command used: /opt/homebrew/opt/coreutils/libexec/gnubin/readlink -->
<p>Activate the new environment to use it:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>conda activate igvtools</code></pre>
</div>
<p>Test that the <code>igv</code> command is available:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>which igv</code></pre>
<pre><code>## /opt/miniconda3/envs/igvtools/bin/igv</code></pre>
</div>
</div>
<div id="igv-videos" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> IGV videos</h2>
<p>The developers of IGV have produced a number of tutorial videos which describe the layout and functionality of the browser. Each video is roughly 5 minutes long and contains a lot of useful information. Instead of needlessly creating a new tutorial, we suggest you watch each of the videos instead.</p>
<div id="data-navigation" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Data navigation</h3>
<p>This video demonstrates how to navigate the browser:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/YpNg0hNUuo8" width="830" height="467" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
<div id="sequencing-data" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Sequencing data</h3>
<p>This video demonstrates how to load sequencing data:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/E_G8z_2gTYM" width="830" height="467" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
<div id="genomic-variation" class="section level3" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Genomic variation</h3>
<p>This video demonstrates how SNPs and indels are displayed:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/eZUI6lcIX3I" width="830" height="467" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
<div id="rna-sequencing" class="section level3" number="3.2.4">
<h3><span class="header-section-number">3.2.4</span> RNA sequencing</h3>
<p>This video demonstrates how RNA-seq data is displayed:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/awGN-rpLYas" width="830" height="467" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
<div id="variation-calling" class="section level3" number="3.2.5">
<h3><span class="header-section-number">3.2.5</span> Variation calling</h3>
<p>This video demonstrates how variant calls are displayed:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/EpD2ZHM7Q8Q" width="830" height="467" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>
</div>
</div>
</div>
<div id="batch-commands" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Batch commands</h2>
<p>In some cases it is useful to control IGV programmatically, rather than interactively. This is handy when you want to perform lots of tasks in the session without having to manually load and navigate the browser. Tasks are issued using a batch script, a text file containing commands which the browser understands. The commands are run sequentially and appear on separate lines.</p>
<p>Below is an example of a batch script. The script starts by creating a new session, performing some tasks, and then exiting the session:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>cat tutorial/data/example/script.txt</code></pre>
<pre><code>## new
## genome hg19
## goto chr12:7,939,997-7,953,742
## snapshot tutorial/data/example/snapshot.png
## exit</code></pre>
</div>
<p>The script is run by passing it to the IGV launcher with the batch parameter:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<pre class="bash"><code>igv --batch tutorial/data/example/script.txt</code></pre>
<pre><code>## Using system JDK.
## WARNING: package com.sun.java.swing.plaf.windows not in java.desktop
## WARNING: package sun.awt.windows not in java.desktop
## openjdk version &quot;17.0.3&quot; 2022-04-19 LTS
## OpenJDK Runtime Environment Zulu17.34+19-CA (build 17.0.3+7-LTS)
## OpenJDK 64-Bit Server VM Zulu17.34+19-CA (build 17.0.3+7-LTS, mixed mode, sharing)
## INFO [Sept 21,2022 10:54] [Main] Startup  IGV Version user not_set
## INFO [Sept 21,2022 10:54] [Main] Java 17.0.3 (build 17.0.3+7-LTS) 2022-04-19
## INFO [Sept 21,2022 10:54] [Main] Java Vendor: Azul Systems, Inc. http://www.azul.com/
## INFO [Sept 21,2022 10:54] [Main] JVM: OpenJDK 64-Bit Server VM Zulu17.34+19-CA   
## INFO [Sept 21,2022 10:54] [Main] OS: Mac OS X 12.6 x86_64
## INFO [Sept 21,2022 10:54] [Main] IGV Directory: /Users/James/igv
## INFO [Sept 21,2022 10:54] [Main] Resoluction scale = 0.0
## SLF4J: Failed to load class &quot;org.slf4j.impl.StaticLoggerBinder&quot;.
## SLF4J: Defaulting to no-operation (NOP) logger implementation
## SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
## INFO [Sept 21,2022 10:54] [AmazonUtils] AWS default credentials found. AWS support enabled.
## INFO [Sept 21,2022 10:54] [CommandListener] Listening on port 60151
## INFO [Sept 21,2022 10:54] [BatchRunner] Executing batch script: tutorial/data/example/script.txt
## INFO [Sept 21,2022 10:54] [GenomeManager] Loading genome: https://s3.amazonaws.com/igv.org.genomes/hg19/hg19.json
## INFO [Sept 21,2022 10:54] [TrackLoader] Loading resource:  https://s3.amazonaws.com/igv.org.genomes/hg19/ncbiRefSeq.sorted.txt.gz
## INFO [Sept 21,2022 10:54] [ShutdownThread] Shutting down</code></pre>
</div>
<p>The browser outputs a picture of the Nanog locus in the human genome:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<img src="data/visualization/chunks/example/snapshot.png" width="100%" style="display: block; margin: auto;" />
</div>
<p>This example is fairly simple, but more complicated tasks can be achieved with additional commands. The full list of batch commands is displayed below.</p>
</div>
<div id="reference-sheet" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Reference sheet</h2>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<table>
<colgroup>
<col width="3%" />
<col width="13%" />
<col width="83%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Command</th>
<th align="left">Parameters</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">collapse</td>
<td align="left">trackName</td>
<td align="left">Collapses a given track. trackName is optional, if it is not supplied all tracks are collapsed.</td>
</tr>
<tr class="even">
<td align="left">colorBy</td>
<td align="left">option tagName</td>
<td align="left">Sets the color by option for alignment tracks. For option TAG also specify the tag name.</td>
</tr>
<tr class="odd">
<td align="left">echo</td>
<td align="left">Writes echo back to the response, primarily for testing port connections.</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">exit</td>
<td align="left">Exit (close) the IGV application.</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">expand</td>
<td align="left">trackName</td>
<td align="left">Expands a given trackName. trackName is optional, however, and if it is not supplied all tracks are expanded.</td>
</tr>
<tr class="even">
<td align="left">genome</td>
<td align="left">genomeIdOrPath</td>
<td align="left">Selects a genome by id, or loads a genome (or indexed fasta) from the supplied path.</td>
</tr>
<tr class="odd">
<td align="left">goto</td>
<td align="left">locus or listOfLoci</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">group</td>
<td align="left">option tagName</td>
<td align="left">Alignment tracks only. Group alignments by the specified option. See below for valid option values. For option TAG also specify the tag name.</td>
</tr>
<tr class="odd">
<td align="left">load</td>
<td align="left">file</td>
<td align="left">Loads a data or session file by specifying a full path to a local file or a URL. To explicitly specify a path to an index file use the optional index= parameter. For examples load foo.bam index=bar.bai</td>
</tr>
<tr class="even">
<td align="left">maxPanelHeight</td>
<td align="left">height</td>
<td align="left">Sets the number of vertical pixels (height) of each panel to include in image. Images created from a port command or batch script are not limited to the data visible on the screen. Stated another way, images can include the entire panel not just the portion visible in the scrollable screen area. The default value for this setting is 1000, increase it to see more data, decrease it to create smaller images. To capture the exact area visible on the screen set this value to -1.</td>
</tr>
<tr class="odd">
<td align="left">new</td>
<td align="left">Create a new session. Unloads all tracks except the default genome annotations.</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">preference</td>
<td align="left">key value</td>
<td align="left">Temporarily set the preference named key to the specified value. This preference only lasts until IGV is shut down. The complete set of preference keys are listed in the file preferences.tab here. The first column is the key, the third column is the value type. For select value types the permitted values follow as a list delimited by the character |.</td>
</tr>
<tr class="odd">
<td align="left">region</td>
<td align="left">chr start end</td>
<td align="left">Defines a region of interest bounded by the two loci (e.g., region chr1 100 200).</td>
</tr>
<tr class="even">
<td align="left">saveSession</td>
<td align="left">filename</td>
<td align="left">Save the current session. It is recommended that a full path be used for filename. IGV release 2.11.1</td>
</tr>
<tr class="odd">
<td align="left">setAltColor</td>
<td align="left">colorString trackName</td>
<td align="left">Set the track altColor, used for negative values in a wig track or negative strand features. See description of setColor below. IGV release 2.11.1</td>
</tr>
<tr class="even">
<td align="left">setColor</td>
<td align="left">colorString trackName</td>
<td align="left">Set the track color. colorString can be a comma delimited rgb string with components in the range 0-255, for example 255,0,0, or a hex color string, for example FF0000. IGV release 2.11.1</td>
</tr>
<tr class="odd">
<td align="left">setDataRange</td>
<td align="left">rangeString trackName</td>
<td align="left">Set the data range (scale) for all numeric tracks, or if a trackName is specified a specific track. rangeString is either a 2 comma delimited list for min,max. As of release 2.11.0 ‘auto’ can be used for rangeString, which will set the track(s) to autoscale.</td>
</tr>
<tr class="even">
<td align="left">setLogScale</td>
<td align="left">true/false trackName</td>
<td align="left">Set the data scale to log (true) or linear (false). Optionally specify a track, if no track is specified all numeric tracks will be set.</td>
</tr>
<tr class="odd">
<td align="left">setSleepInterval</td>
<td align="left">ms</td>
<td align="left">Sets a delay (sleep) time in milliseconds. The sleep interval is invoked between successive commands.</td>
</tr>
<tr class="even">
<td align="left">setTrackHeight</td>
<td align="left">height trackName</td>
<td align="left">Set the specified track’s height in integer units. trackName is required.</td>
</tr>
<tr class="odd">
<td align="left">snapshotDirectory</td>
<td align="left">path</td>
<td align="left">Sets the directory in which to write images.</td>
</tr>
<tr class="even">
<td align="left">snapshot</td>
<td align="left">filename</td>
<td align="left">Saves a snapshot of the IGV window to an image file. If filename is omitted, writes a PNG file with a filename generated based on the locus. If filename is specified, the filename extension determines the image file format, which must be either .png or .svg.</td>
</tr>
<tr class="odd">
<td align="left">sort</td>
<td align="left">option locus</td>
<td align="left">Sorts alignment or segmented copy number tracks. See below for valid option values. If supplied, the locus option can define a single position, or a range. If absent sorting will be based on the region in view for segmented copy number, or the center position of the region in view for alignment tracks.</td>
</tr>
<tr class="even">
<td align="left">squish</td>
<td align="left">trackName</td>
<td align="left">Squish a given trackName. trackName is optional, and if it is not supplied all annotation tracks are squished.</td>
</tr>
<tr class="odd">
<td align="left">viewaspairs</td>
<td align="left">trackName</td>
<td align="left">Set the display mode for an alignment track to View as pairs. trackName is optional.</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="exercises" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Exercises</h1>
<p>The exercises below are designed to strengthen your knowledge of using IGV and writing batch commands. The solution to each problem is blurred, only after attempting to solve the problem yourself should you look at the solution. Should you need any help, please ask one of the instructors.</p>
<div id="protein-dna-binding" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Protein-DNA binding</h2>
<blockquote>
<p>ChIP-sequencing, also known as ChIP-seq, is a method used to analyze protein interactions with DNA. ChIP-seq combines chromatin immunoprecipitation (ChIP) with massively parallel DNA sequencing to identify the binding sites of DNA-associated proteins. It can be used to map global binding sites precisely for any protein of interest.</p>
</blockquote>
<p>In the <code>chipseq</code> directory are a set of files generated from the analysis of a ChIP-seq experiment:</p>
<ul>
<li>ZHBTC4_OCT4_UNT.bw - Coverage track for the DNA-binding protein OCT4</li>
<li>ZHBTC4_Input_UNT.bw - Coverage track for the input genomic DNA</li>
<li>ZHBTC4_OCT4_UNT_peaks.narrowPeak - Peak regions detected by MACS2 peak caller</li>
<li>ZHBTC4_OCT4_UNT_summits.bed - Summit regions detected by MACS2 peak caller</li>
</ul>
<p>All of the data files are compatible with the <strong>mm10</strong> mouse reference genome.</p>
<p>Using both the IGV browser and the command line, answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>Which peak has the highest score? Type <code>man sort</code> for help sorting.</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="bash"><code># ZHBTC4_OCT4_UNT_peak_24
sort -k5,5nr tutorial/data/chipseq/ZHBTC4_OCT4_UNT_peaks.narrowPeak | head -n 1</code></pre>
<pre><code>## chr19    7261376 7262037 ZHBTC4_OCT4_UNT_peak_24 1690    .   36.5612 176.877 169.089 326</code></pre>
</div>
</div>
</div>
<ol start="2" style="list-style-type: decimal">
<li>What gene is nearest to the highest scoring peak?</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># Rcor2</code></pre>
<img src="data/visualization/chunks/chipseq/script1.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
<ol start="3" style="list-style-type: decimal">
<li>Which peak is the longest? Type <code>man awk</code> for help calculating.</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
bash
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="bash"><code># ZHBTC4_OCT4_UNT_peak_149
awk &#39;{print $1, $2, $3, $4, $3 - $2}&#39; tutorial/data/chipseq/ZHBTC4_OCT4_UNT_peaks.narrowPeak | sort -k5,5nr | head -n 1</code></pre>
<pre><code>## chr19 38347912 38349082 ZHBTC4_OCT4_UNT_peak_149 1170</code></pre>
</div>
</div>
</div>
<ol start="4" style="list-style-type: decimal">
<li>What base is covered by the summit of the longest peak?</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># G</code></pre>
<img src="data/visualization/chunks/chipseq/script2.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
<ol start="5" style="list-style-type: decimal">
<li>How many peaks are located inside the Cep55 gene?</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># 3</code></pre>
<img src="data/visualization/chunks/chipseq/script3.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
</div>
<div id="genomic-variation-1" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Genomic variation</h2>
<blockquote>
<p>Exome sequencing, also known as whole exome sequencing (WES), is a genomic technique for sequencing all of the protein-coding regions of genes in a genome (known as the exome). It consists of two steps: the first step is to select only the subset of DNA that encodes proteins. These regions are known as exons. The second step is to sequence the exonic DNA using any high-throughput DNA sequencing technology.</p>
</blockquote>
<p>In the <code>exome</code> directory is a VCF file generated from the analysis of a whole-exome sequencing (WES) experiment. The file is called <code>1KGP.vcf</code> and contains variant calling information from a number of samples from the <a href="https://www.internationalgenome.org">1000 Genomes Project</a> data portal. The VCF file is compatible with the <strong>hg39</strong> human reference genome.</p>
<p>Load the VCF file in the IGV browser and answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>How many samples are represented in the VCF file?</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># 10</code></pre>
<img src="data/visualization/chunks/exome/q1.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
<ol start="2" style="list-style-type: decimal">
<li>What is the reference and alternate allele at position chr19:4196493</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># Reference: A
# Alternate: G</code></pre>
<img src="data/visualization/chunks/exome/q2.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
<ol start="3" style="list-style-type: decimal">
<li>How many SNPs are present in the ACP7 gene?</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># 1
# The first SNP is homozygous reference</code></pre>
<img src="data/visualization/chunks/exome/q3.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
<ol start="4" style="list-style-type: decimal">
<li>What two positions show homozygous variants in all 10 samples?</li>
</ol>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># chr19:9387798
# chr19:55999276</code></pre>
<img src="data/visualization/chunks/exome/q4.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
<ol start="5" style="list-style-type: decimal">
<li>How many of the following alleles are present at position chr19:4355511</li>
</ol>
<ul>
<li>Homozygous reference</li>
<li>Homozygous variant</li>
<li>Heterozygous variant</li>
</ul>
<div class="chunk-title-parent">
<div class="chunk-title small">
R
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="r"><code># Homozygous reference: 1
# Homozygous variant: 3
# Heterozygous variant: 6</code></pre>
<img src="data/visualization/chunks/exome/q5.png" width="100%" style="display: block; margin: auto;" />
</div>
</div>
</div>
</div>
<div id="gene-expression" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Gene expression</h2>
<blockquote>
<p>RNA-Seq (named as an abbreviation of RNA sequencing) is a sequencing technique which uses next-generation sequencing (NGS) to reveal the presence and quantity of RNA in a biological sample at a given moment, analyzing the continuously changing cellular transcriptome.</p>
</blockquote>
<p>In the <code>rnaseq</code> directory are a set of coverage tracks generated from the analysis of an RNA-seq experiment. All of the data files are compatible with the <strong>mm10</strong> mouse reference genome. Write a batch script to recreate the snapshot shown below:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
Output
</div>
<img src="data/visualization/chunks/rnaseq/snapshot.png" width="100%" style="display: block; margin: auto;" />
</div>
<p>Click the bash chunk to reveal the solution:</p>
<div class="chunk-title-parent">
<div class="chunk-title small">
verbatim
</div>
<div class="blur-outer">
<div class="blur-title">

</div>
<div class="blur-content">
<pre class="default"><code>new

genome mm10

goto chr19:8,987,120-9,078,935 chr19:38,052,980-38,076,424 chr19:27,386,697-27,431,908

load tutorial/data/rnaseq/BRG1FL_TAM.bw
setColor #FF0000 BRG1FL_TAM.bw

load tutorial/data/rnaseq/BRG1FL_UNT.bw
setColor #0000FF BRG1FL_UNT.bw

load tutorial/data/rnaseq/ZHBTC4_DOX.bw
setColor #008000 ZHBTC4_DOX.bw

load tutorial/data/rnaseq/ZHBTC4_UNT.bw
setColor #800080 ZHBTC4_UNT.bw

setDataRange 0,50

snapshot exercises/expression/snapshot.png

exit</code></pre>
</div>
</div>
</div>
</div>
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">

<div class="footer-left col-sm-6">
<p class="small">
<b>2022</b> • <a href="https://www.zifornd.com">Zifo</a>
</p>
</div>

<div class="footer-right col-sm-6">
<p>
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://www.zifornd.com">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/zifornd"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/zifo-technologies">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
