---
title: "SAM Format- Samtools"
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r,child="assets/header-lab.Rmd"}
```

```{r, include = FALSE}
knitr::opts_chunk$set(chunk.title = TRUE, engine.opts = list(bash = "-l"))
```

# Introduction

## What is the SAM file format?

From Wikipedia, the free encyclopedia:

> Sequence Alignment Map (SAM) is a text-based format originally for storing biological sequences aligned to a reference sequence developed by Heng Li and Bob Handsaker et al.[1] It was developed when the 1000 Genomes Project wanted to move away from the MAQ mapper format and decided to design a new format. The overall TAB-delimited flavour of the format came from an earlier format inspired by BLATâ€™s PSL. The name of SAM came from Gabor Marth from University of Utah, who originally had a format under the same name but with a different syntax more similar to a BLAST output.[2] It is widely used for storing data, such as nucleotide sequences, generated by next generation sequencing technologies, and the standard has been broadened to include unmapped sequences.[3] The format supports short and long reads (up to 128 Mbp) produced by different sequencing platforms and is used to hold mapped data within the Genome Analysis Toolkit (GATK) and across the Broad Institute, the Wellcome Sanger Institute, and throughout the 1000 Genomes Project.

# What is the Samtools software?

From Wikipedia, the free encyclopedia:

> SAMtools is a set of utilities for interacting with and post-processing short DNA sequence read alignments in the SAM (Sequence Alignment/Map), BAM (Binary Alignment/Map) and CRAM formats, written by Heng Li. These files are generated as output by short read aligners like BWA. Both simple and advanced tools are provided, supporting complex tasks like variant calling and alignment viewing as well as sorting, indexing, data extraction and format conversion.[3] SAM files can be very large (10s of Gigabytes is common), so compression is used to save space. SAM files are human-readable text files, and BAM files are simply their binary equivalent, whilst CRAM files are a restructured column-oriented binary container format. BAM files are typically compressed and more efficient for software to work with than SAM. SAMtools makes it possible to work directly with a compressed BAM file, without having to uncompress the whole file. Additionally, since the format for a SAM/BAM file is somewhat complex - containing reads, references, alignments, quality information, and user-specified annotations - SAMtools reduces the effort needed to use SAM/BAM files by hiding low-level details.

# Objectives

* [Software installation]
* [Data preparation]
* [SAM reading/writing]
* [Alignment statistics]
* [Alignment filtering]

# Tutorial

The first thing to do is create a directory to store all the tutorial output files and data:

```{bash, echo = -1}
rm -rf tutorial
mkdir tutorial
```

## Software installation

Create a new environment called 'alignment' with all the required software installed:

```{bash}
conda create --yes --name alignment bowtie2 bwa entrez-direct samtools seqkit sra-tools=2.11.0 # only this version works currently (21/09/2022)
```

<!-- Download xtract for Apple Silicon -->

```{bash, include = FALSE}
curl -o /opt/miniconda3/envs/alignment/bin/xtract.Silicon.gz https://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/xtract.Silicon.gz
gunzip /opt/miniconda3/envs/alignment/bin/xtract.Silicon.gz
chmod +x /opt/miniconda3/envs/alignment/bin/xtract.Silicon
```

Activate the 'alignment' environment:

```{bash}
conda activate alignment
```

Test that Bowtie 2 and BWA are available:

```{bash, error = TRUE, echo = -1}
conda activate alignment
which samtools
```

## Data preparation

Download the Ebola virus reference genome:

```{bash, echo = -1}
conda activate alignment
efetch -db nuccore -id AF086833 -format fasta > tutorial/AF086833.fasta
```

Download 1,000 sequencing reads from public data:

```{bash, echo = -1}
conda activate alignment
fastq-dump -X 1000 -O tutorial --split-files SRR1972739
```

Index the reference genome:

```{bash, echo = -1}
conda activate alignment
bwa index -p tutorial/AF086833 tutorial/AF086833.fasta
```

Align reads using BWA to the reference genome:

```{bash, echo = -1}
conda activate alignment
bwa mem tutorial/AF086833 tutorial/SRR1972739_1.fastq tutorial/SRR1972739_2.fastq > tutorial/SRR1972739.sam
```

## SAM reading/writing

The SAM format is used to store output from read alignment software. It is a TAB-delimited text format consisting of a header and alignment section. To look at each of these sections, we can use the `view` command.

Print the SAM header section:

```{bash, echo = -1}
conda activate alignment
samtools view -H tutorial/SRR1972739.sam
```

The header section primarily contains metadata about the file and reference sequence. For example, whether the alignments in the file have been sorted. Refer to the [SAM specification](http://samtools.github.io/hts-specs/SAMv1.pdf) for full details.

Print the SAM alignment section:

```{bash, echo = -1}
conda activate alignment
# Output the first few lines
samtools view tutorial/SRR1972739.sam | head
```

The alignment section contains data about the input sequence alignments. Each line typically represents a single alignment and contains information about how the alignment was generated. For example, which reference sequence and at what position was the input sequence aligned. Again, refer to the [SAM specification](http://samtools.github.io/hts-specs/SAMv1.pdf) for full details.

It's important to note that we rarely work with alignments in this text-based format. Usually, we convert the SAM file into a binary or compressed representation which we refer to as the BAM format. This allows output from large sequencing runs to be stored efficiently. It also helps if we sort the alignments by their mapping position as this allows us to produce what's called an 'index' file. As the name suggests, this file is analogous to the index in a book. It tells us which part of the file to read to fetch alignments from a particular region of the reference sequence. To sort the alignments and output a BAM file, we can use the `sort` command.

Sort and convert from SAM to BAM format:

```{bash, echo = -1}
conda activate alignment
samtools sort tutorial/SRR1972739.sam > tutorial/SRR1972739.bam
```

You should be aware that older versions of `samtools` required two commands to achieve the same outcome:

```{bash, eval = FALSE}
samtools view -b tutorial/SRR1972739.sam > tutorial/SRR1972739.bam
samtools sort tutorial/SRR1972739.bam > tutorial/SRR1972739.sorted.bam
```

Once we have the sorted BAM file, we can then create an index file. We do this using the `index` command. Notably, we do not have to specify an output file name for this command. The index file is automatically created and named by appending `.bai` to the input file name.

Build an index for the BAM file:

```{bash, echo = -1}
conda activate alignment
samtools index tutorial/SRR1972739.bam
```

List the BAM and index file:

```{bash, echo = -1}
conda activate alignment
ls tutorial/SRR1972739.*
```

## Alignment statistics

It's often useful to calculate various statistics on alignment files. These statistics can be used to answer a range of questions, some needed for basic quality control and others which have biological significance. We will go through the most common statistics which can be calculated by the `samtools` commands. Alignment summary statistics can be reported using the `idxstats` command. The output is TAB-delimited with each line consisting of reference sequence name, sequence length, number of mapped reads, and number of unmapped reads.

```{bash, echo = -1}
conda activate alignment
samtools idxstats tutorial/SRR1972739.bam
```

In addition to reporting the number of alignments by reference sequence, we can use the `flagstat` command to report the number of alignments for each FLAG value as described in the SAM format specification.

```{bash, echo = -1}
conda activate alignment
samtools flagstat tutorial/SRR1972739.bam
```

The `flagstat` report is helpful to assess the overall alignment performance. For example, a low mapping rate may indicate a significant problem with the library and will require further investigation. If you ever find yourself confused about what a FLAG value represents, you can use the `flags` command to convert between the value and it's description.

Convert FLAG value to description:

```{bash, echo = -1}
conda activate alignment
samtools flags PAIRED
```

Convert description to FLAG value:

```{bash, echo = -1}
conda activate alignment
samtools flags 1
```

In many genomics applications it is important to compute both read coverage and read depth at each position or region:

  - Read coverage is defined as the proportion of a region which is covered by alignments. For example, a set of alignments might cover only 10% of the full reference sequence.
  - Read depth is defined as the number of alignments that cover a particular position. For example, the 1st base in the reference sequence may be covered by 100 alignments.

To calculate read coverage, we can use the `coverage` command. The output contains some useful values like the number of covered bases (covbases), the percentage of covered bases (coverage), and the mean depth of coverage (meandepth).

```{bash, echo = -1}
conda activate alignment
samtools coverage tutorial/SRR1972739.bam
```

To calculate read depth, we can use the `depth` command. The output is TAB-delimited with each line consisting of reference sequence name, coordinate, and number of mapped reads or alignments.

```{bash, echo = -1}
conda activate alignment
# Output the first few lines
samtools depth tutorial/SRR1972739.bam | head
```

## Alignment filtering

For many genomics applications, the SAM file produced by most read alignment software will require some additional processing. You might have to remove unmapped reads, select only high-quality alignments, or use just a subset of regions. All of these tasks and more can be accomplished using the `view` command. There are a number of options available in this command, so it is worth examining the help information first:

```{bash, echo = -1}
conda activate alignment
samtools view --help
```

As you can see, there are a lot of options. These allow the user to perform a huge variety of filtering and file manipulation. These options become especially powerful when combined with other `samtools` commands. Rather than run through an exhaustive tour of the `view` command, we will instead demonstrate the most typical use cases.

To filter alignments, we can use the `-f` and `-F` options. The first option will output alignments *with* a given FLAG value, and the second will output alignments *without* a given FLAG value. This can be quite counter-intuitive so we should look at some examples:

Output alignments where the read is unmapped:

```{bash, echo = -1}
conda activate alignment
# UNMAP (4)
# With the UNMAP flag (-f 4)
samtools view -f 4 tutorial/SRR1972739.bam > tutorial/unmapped.sam
```

Output alignments where the read is mapped (not unmapped):

```{bash, echo = -1}
conda activate alignment
# UNMAP (4)
# Without the UNMAP flag (-F 4)
samtools view -F 4 tutorial/SRR1972739.bam > tutorial/mapped.bam
```

Remember, that FLAG values can be summed together to represent a combination of FLAG values.

Output alignments where the reads is unmapped and is the first read of the pair:

```{bash, echo = -1}
conda activate alignment
# UNMAP (4)
# READ1 (64)
# UNMAP,READ1 (4 + 64 = 68)
# With the UNMAP,READ1 flag combination (-f 68)
samtools view -f 68 tutorial/SRR1972739.bam > tutorial/unmapped_read1.bam
```

Output alignments where the reads is mapped (not unmapped) and passed quality control (did not fail quality control):

```{bash, echo = -1}
conda activate alignment
# UNMAP (4)
# QCFAIL (512)
# UNMAP,QCFAIL (4 + 512 = 516)
# Without the UNMAP,QCFAIL flag combination (-F 516)
samtools view -F 516 tutorial/SRR1972739.bam > tutorial/mapped_qc.bam
```

To perform more specific filtering - and to make matters more confusing - you can select and disregard combinations of FLAG values. For example, to output alignments on the reverse strand we need to identify alignments *without* the unmapped FLAG value and *with* the reverse complemented FLAG value:

```{bash, echo = -1}
conda activate alignment
samtools view -c -F 4 -f 16 tutorial/SRR1972739.bam > tutorial/mapped_reverse.bam
```

Do not be discouraged if you find the FLAG values and the *with* or *with not* system of alignment output difficult to remember. Bioinformaticians the world over look this information up on a daily basis and there is no shame in searching online for a solution or reference. Aside from using FLAG values to output specific alignments, we can also choose based on the mapping quality. We can use the `-q` option to output alignments with mapping quality greater than or equal to a desired value:

```{bash, echo = -1}
conda activate alignment
# mapping quality >= 10
samtools view -b -q 10 tutorial/SRR1972739.bam > tutorial/SRR1972739.mapq.bam
```

Lastly, we can also output alignments from a specific region of the reference sequence. This is useful when you may want to focus on alignments which cover a particular region of the genome, like a specific gene body. To output alignments for a given region, simply specify the region on the command line with the required format (NAME:START-END).

```{bash, echo = -1}
conda activate alignment
# NAME = AF086833
# START = 1
# END = 1000
samtools view -b tutorial/SRR1972739.bam AF086833.2:1-1000 > tutorial/SRR1972739.region.bam
```

Finally, Samtools also provides a text-based genome browser to visually inspect the alignments:

```{bash, echo = -1}
conda activate alignment
# display as text (-d T)
# go to this position (-p AF086833.2:9972)
samtools tview -d T -p AF086833.2:9972 tutorial/SRR1972739.bam tutorial/AF086833.fasta
```

# Exercise

The exercises below are designed to strengthen your knowledge of using Samtools to manipulate SAM files. The solution to each problem is blurred, only after attempting to solve the problem yourself should you look at the solution. Should you need any help, please ask one of the instructors.

Create a directory to store the output files from each exercise:

```{bash, echo = -1}
rm -rf exercises
mkdir exercises
mkdir exercises/cholera
mkdir exercises/zika
mkdir exercises/tuberculosis
```

## Cholera

> Cholera is an acute diarrhoeal infection caused by eating or drinking food or water that is contaminated with the bacterium Vibrio cholerae. Cholera remains a global threat to public health and is an indicator of inequity and lack of social development. Researchers have estimated that every year, there are 1.3 to 4.0 million cases of cholera, and 21,000 to 143,000 deaths worldwide due to the infection.
> `r tufte::quote_footer('World Health Organization')`

1. Use the `efetch` command to download the genome for the Vibrio cholerae bacteria. Note, the genome is split into two entries on the NCBI database: AP014524 and AP014525. You can either download each seperately and then concatenate the files or provide both accession numbers to the `efetch` command.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
efetch -db nuccore -id "AP014524,AP014525" -format fasta > exercises/cholera/MS6.fasta
```

2. Use the `fastq-dump` command to download 1000 single-end reads from the public Vibrio cholerae sequencing run with accession number SRR16345277.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
fastq-dump -X 1000 -O exercises/cholera SRR16345277
```

3. Index the bacterial genome using BWA

```{bash, blur = FALSE, echo = -1}
conda activate alignment
bwa index exercises/cholera/MS6.fasta
```

4. Align the sequencing reads to the bacterial genome using BWA

```{bash, blur = FALSE, echo = -1}
conda activate alignment
bwa mem exercises/cholera/MS6.fasta exercises/cholera/SRR16345277.fastq > exercises/cholera/SRR16345277.sam
```

5. Use Samtools to sort and convert the SAM file to BAM format

```{bash, blur = FALSE, echo = -1}
conda activate alignment
samtools sort exercises/cholera/SRR16345277.sam > exercises/cholera/SRR16345277.bam
```

6. Index the sorted BAM file

```{bash, blur = FALSE, echo = -1}
conda activate alignment
samtools index exercises/cholera/SRR16345277.bam
```

7. Calculate flag alignment statistics from the BAM file

```{bash, blur = FALSE, echo = -1}
conda activate alignment
samtools flagstat exercises/cholera/SRR16345277.bam
```

8. Calculate index alignment statistics from the BAM file

```{bash, blur = FALSE, echo = -1}
conda activate alignment
samtools idxstats exercises/cholera/SRR16345277.bam
```

9. Filter the BAM file to contain only MAPPED reads

```{bash, blur = FALSE, echo = -1}
conda activate alignment
# UNMAP (4)
samtools view -b -F 4 exercises/cholera/SRR16345277.bam > exercises/cholera/SRR16345277.mapped.bam
```

## Zika

> Zika virus is primarily transmitted by the bite of an infected mosquito from the Aedes genus, mainly Aedes aegypti, in tropical and subtropical regions. Aedes mosquitoes usually bite during the day, peaking during early morning and late afternoon/evening. This is the same mosquito that transmits dengue, chikungunya and yellow fever.
> `r tufte::quote_footer('World Health Organization')`

1. Use the `efetch` command from EDirect to download and save the Zika reference genome (AY632535).

```{bash, blur = FALSE, echo = -1}
conda activate alignment
efetch -db nuccore -id AY632535 -format fasta > exercises/zika/AY632535.fasta
```

2. Use the `esearch` and `efetch` commands to download and save the run information from a public Zika sequencing project (PRJNA609110).

```{bash, eval = FALSE, blur = TRUE, echo = -1}
conda activate alignment
esearch -db sra -query PRJNA609110 | efetch -format runinfo > exercises/zika/runinfo.csv
```

<!-- The esearch command only works asynchronously -->

```{bash, include = FALSE}
conda activate alignment
esearch -db sra -query PRJNA609110 > /tmp/query.txt &
sleep 10
cat /tmp/query.txt | efetch -format runinfo > exercises/zika/runinfo.csv
```

3. Create a file which contains the run accession numbers of those runs which are single-end.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
grep "SINGLE" exercises/zika/runinfo.csv | cut -d "," -f 1 > exercises/zika/runids.txt
```

4. Use the `fastq-dump` command to download 10,000 single-end reads from each of the runs selected in Q3. Instead of manually creating the command for each run accession number, use a for loop to read each run accession number from the file created in Q3.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  fastq-dump -X 10000 -O exercises/zika ${RUN}

done < exercises/zika/runids.txt
```

5. Use the BWA aligner to index the Zika genome:

```{bash, blur = FALSE, echo = -1}
conda activate alignment
bwa index exercises/zika/AY632535.fasta
```

6. Align the sequencing reads to the Zika genome use the BWA aligner. Again, use a for loop to construct and run each of the commands.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  bwa mem exercises/zika/AY632535.fasta exercises/zika/${RUN}.fastq > exercises/zika/${RUN}.sam

done < exercises/zika/runids.txt
```

7. Sort and convert the SAM files to BAM format. Use a loop as before.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  samtools sort exercises/zika/${RUN}.sam > exercises/zika/${RUN}.bam

done < exercises/zika/runids.txt
```

8. Index the sorted BAM files. Loop again.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  samtools index exercises/zika/${RUN}.bam

done < exercises/zika/runids.txt
```

9. Use the `merge` command to combine the alignments from all of the BAM files into a single BAM file.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
samtools merge exercises/zika/merge.bam exercises/zika/SRR*.bam
```

10. Calculate the read coverage from the merged BAM file and report the percent covered and mean coverage values.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
# Percent covered: 5.151%
# Mean coverage: 238x
samtools coverage -m exercises/zika/merge.bam
```

## Tuberculosis

> Tuberculosis (TB) is caused by bacteria (Mycobacterium tuberculosis) and it most often affects the lungs. TB is spread through the air when people with lung TB cough, sneeze or spit. A person needs to inhale only a few germs to become infected. Every year, 10 million people fall ill with tuberculosis (TB). Despite being a preventable and curable disease, 1.5 million people die from TB each year â€“ making it the worldâ€™s top infectious killer.
> `r tufte::quote_footer('World Health Organization')`

1. Use the `efetch` command from EDirect to download and save the Tuberculosis reference genome (AL123456).

```{bash, blur = FALSE, echo = -1}
conda activate alignment
efetch -db nuccore -id AL123456 -format fasta > exercises/tuberculosis/AL123456.fasta
```

2. Use the `esearch` and `efetch` commands to download and save the run information from a public Tuberculosis sequencing project (PRJNA575883).

```{bash, eval = FALSE, blur = TRUE, echo = -1}
conda activate alignment
esearch -db sra -query PRJNA575883 | efetch -format runinfo > exercises/tuberculosis/runinfo.csv
```

<!-- The esearch command only works asynchronously -->

```{bash, include = FALSE}
conda activate alignment
esearch -db sra -query PRJNA575883 > /tmp/query.txt &
sleep 10
cat /tmp/query.txt | efetch -format runinfo > exercises/tuberculosis/runinfo.csv
```

3. Create a file which contains the run accession numbers of those runs from submission SRA1348739.

```{bash, blur = TRUE, echo = -1}
conda activate alignment
cat exercises/tuberculosis/runinfo.csv | grep "SRA1348739" | cut -d "," -f 1 > exercises/tuberculosis/runids.txt
```

4. Use the `fastq-dump` command to download 1,000 paired-end reads from each of the runs selected in Q3. Instead of manually creating the command for each run accession number, use a for loop to read each run accession number from the file created in Q3.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  fastq-dump -X 1000 -O exercises/tuberculosis --split-files ${RUN}

done < exercises/tuberculosis/runids.txt
```

5. Use the Bowtie 2 aligner to index the Tuberculosis genome.

```{bash, blur = TRUE, echo = -1}
conda activate alignment
bowtie2-build exercises/tuberculosis/AL123456.fasta exercises/tuberculosis/AL123456
```

6. Align the sequencing reads to the Tuberculosis genome use the Bowtie 2 aligner. Additionally, sort, convert, and index the SAM file. Again, use a for loop to construct and run each of the commands.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  bowtie2 -x exercises/tuberculosis/AL123456 -1 exercises/tuberculosis/${RUN}_1.fastq -2 exercises/tuberculosis/${RUN}_2.fastq > exercises/tuberculosis/${RUN}.sam
  
  samtools sort exercises/tuberculosis/${RUN}.sam > exercises/tuberculosis/${RUN}.bam
  
  samtools index exercises/tuberculosis/${RUN}.bam

done < exercises/tuberculosis/runids.txt
```

7. Filter the BAM files so they contain only properly paired alignments. Remember to index the new BAM file as well.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
while read RUN; do

  # Output BAM (-b)

  # FLAGs present (PROPER_PAIR = 2)

  samtools view -b -f 2 exercises/tuberculosis/${RUN}.bam > exercises/tuberculosis/${RUN}.PROPER_PAIR.bam
  
  samtools index exercises/tuberculosis/${RUN}.PROPER_PAIR.bam

done < exercises/tuberculosis/runids.txt
```

8. Calculate the read depth at each position in the genome across all BAM files into a file.

```{bash, blur = FALSE, echo = -1}
conda activate alignment
samtools depth -H exercises/tuberculosis/*.PROPER_PAIR.bam > exercises/tuberculosis/depth.txt
```

<!-- Clean up conda and files -->

```{bash, include = FALSE}
rm -rf tutorial
rm -rf exercises
conda remove --name alignment --all --yes
```

```{r,child="assets/footer-lab.Rmd"}
```
