<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Sequence Alignment Map</title>
    <meta charset="utf-8" />
    <meta name="author" content="James Ashmore • 20-Sep-2022" />
    <script src="libs/header-attrs-2.16/header-attrs.js"></script>
    <link href="libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <script src="libs/kePrint-0.0.1/kePrint.js"></script>
    <link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
    <link rel="stylesheet" href="assets/slide.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Sequence Alignment Map
]
.author[
### <b>James Ashmore</b> • 20-Sep-2022
]
.institute[
### Zifo RnD Solutions
]

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;
&lt;link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous"&gt;





&lt;!-- ------------ Only edit title, subtitle &amp; author above this ------------ --&gt;





---

# The SAM format

## Overview

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.sam.png" alt="Fig 1: Example of SAM file format" width="100%" /&gt;
&lt;p class="caption"&gt;Fig 1: Example of SAM file format&lt;/p&gt;
&lt;/div&gt;

* SAM stands for **S**equence **A**lignment **M**ap format

* Store biological sequences aligned to a reference sequence

* TAB-delimited text format consisting of a **header** and **alignment** section

* Each alignment has **11** mandatory fields for essential alignment information

* The full SAM format specification is available [online](http://samtools.github.io/hts-specs/SAMv1.pdf)

* Developed by [Heng Li](https://www.broadinstitute.org/bios/heng-li) and [Bob Handsaker](https://www.broadinstitute.org/bios/bob-handsaker) et al. (2009)

---

# The SAM format

## Header

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.header.png" alt="Fig 2: Example of SAM header" width="100%" /&gt;
&lt;p class="caption"&gt;Fig 2: Example of SAM header&lt;/p&gt;
&lt;/div&gt;

* Each line begins with an `@` character followed by a record type:

  * File-level metadata `HD`
  * Reference sequence dictionary `SQ`
  * Read group `RG`
  * Program `PG`
  * Comment `CO`

* Each record type contains different record tags:

  * For example, `SQ` contains reference sequence name `SN`, length `LN`, etc.

* The full list of record types and their tags is available [online](https://samtools.github.io/hts-specs/SAMv1.pdf#page=3)

---

# The SAM format

## Alignments

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.png" alt="Fig 3: Example of SAM alignments" width="100%" /&gt;
&lt;p class="caption"&gt;Fig 3: Example of SAM alignments&lt;/p&gt;
&lt;/div&gt;

* Each line typically represents the linear alignment of a segment

* Each alignment consists of 11 or more TAB-separated fields:

  * `QNAME` `FLAG` `RNAME` `POS` `MAPQ` `CIGAR` `RNEXT` `PNEXT` `TLEN` `SEQ` `QUAL`

---

# The SAM format

## Alignments

&lt;table class="table" style="font-size: 14px; width: auto !important; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Col &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Field &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Type &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Description &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 1 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; QNAME &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; String &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Query template NAME &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 2 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; FLAG &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Int &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; bitwise FLAG &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 3 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; RNAME &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; String &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Reference sequence NAME &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 4 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; POS &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Int &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 1-based leftmost mapping POSition &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 5 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; MAPQ &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Int &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; MAPping Quality &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 6 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; CIGAR &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; String &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; CIGAR string &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 7 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; RNEXT &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; String &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Reference name of the mate/next read &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 8 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; PNEXT &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Int &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Position of the mate/next read &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 9 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; TLEN &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Int &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; observed Template LENgth &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 10 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; SEQ &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; String &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; segment SEQuence &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 11 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; QUAL &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; String &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; ASCII of Phred-scaled base QUALity+33 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

&lt;center&gt;Tab 1: Alignment fields&lt;/center&gt;

---

# The SAM fields

## QNAME (1)

.pull-left-60[

* Identical QNAME come from the same template

* `*` indicates the information is unavailable

* Reads may occupy multiple alignment lines when:

  * The alignment is chimeric
  * There are multiple mappings

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.QNAME.png" alt="Fig 4: Query template NAME" width="35%" /&gt;
&lt;p class="caption"&gt;Fig 4: Query template NAME&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## FLAG (2)

.pull-left-60[

* The FLAG encodes attributes of a read alignment

* It is displayed as a single integer code:

  * READ MAPPED TO FORWARD STRAND = 0
  * READ UNMAPPED = 4
  * READ MAPPED TO REVERSE STRAND = 16
  * READ FAILED QUALITY CONTROl = 512

* Integers are the sum of **bitwise** flags

* A bitwise flag encodes a specific **attribute**

* The attributes are summed to get the final value

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.FLAG.png" alt="Fig 5: Combination of bitwise FLAGs" width="35%" /&gt;
&lt;p class="caption"&gt;Fig 5: Combination of bitwise FLAGs&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## FLAG (2)

.pull-left-50[

&lt;table class="table" style="font-size: 13px; width: auto !important; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Integer &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Binary &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Description &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 1 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000000000001 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read paired &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 2 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000000000010 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read mapped in proper pair &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 4 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000000000100 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read unmapped &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 8 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000000001000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Mate unmapped &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 16 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000000010000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read reverse strand &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 32 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000000100000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Mate reverse strand &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 64 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000001000000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; First in pair &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 128 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000010000000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Second in pair &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 256 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 000100000000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Not primary alignment &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 512 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 001000000000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read fails platform/vendor quality checks &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 1024 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 010000000000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read is PCR or optical duplicate &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 2048 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 100000000000 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Supplementary alignment &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

&lt;center&gt;Tab 2: bitwise FLAGs&lt;/center&gt;

]

.pull-right-50[


&lt;table class="table" style="font-size: 13px; width: auto !important; margin-left: auto; margin-right: auto;"&gt;
&lt;caption style="font-size: initial !important;"&gt;Ex 1: An umapped read which failed quality control&lt;/caption&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Flag &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Meaning &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Sum &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 4 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read umapped &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 4 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 512 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read fails QC &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 516 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

&lt;table class="table" style="font-size: 13px; width: auto !important; margin-left: auto; margin-right: auto;"&gt;
&lt;caption style="font-size: initial !important;"&gt;Ex 2: A supplementary alignment from the first read in a pair where the mate is aligned to the reverse strand&lt;/caption&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Flag &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Meaning &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Sum &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 1 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Read paired &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 1 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 32 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Mate reverse strand &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 33 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 64 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; First in pair &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 97 &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; 2048 &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Supplementary alignment &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; 2145 &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

]

---

# The SAM fields

## RNAME (3)

.pull-left-60[

* Reference sequence name of the alignment

* An unmapped read has a `*` in this field

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.RNAME.png" alt="Fig 6: Reference sequence NAME" width="35%" /&gt;
&lt;p class="caption"&gt;Fig 6: Reference sequence NAME&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## POS (4)

.pull-left-60[

* 1-based leftmost mapping position of the read

* First base in a reference sequence has coordinate 1

* POS is set as 0 for an unmapped read without coordinate

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.POS.png" alt="Fig 7: 1-based leftmost mapping POSition" width="35%" /&gt;
&lt;p class="caption"&gt;Fig 7: 1-based leftmost mapping POSition&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## MAPQ (5)

.pull-left-60[

* Integer mapping of the probability of an **incorrect** mapping position

* MAPQ calculation: `\(Q=-10\log_{{10}}P\)`

&lt;br&gt;

&lt;table class="table" style="font-size: 15px; margin-left: auto; margin-right: auto;"&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:right;"&gt; Mapping quality &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; Probability of incorrect mapping position &lt;/th&gt;
   &lt;th style="text-align:right;"&gt; Mapping position accuracy &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 10 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 in 10 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 90% &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 20 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 in 100 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 99% &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 30 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 in 1000 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 99.9% &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 40 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 in 10,000 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 99.99% &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 50 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 in 100,000 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 99.999% &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:right;"&gt; 60 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 1 in 1,000,000 &lt;/td&gt;
   &lt;td style="text-align:right;"&gt; 99.9999% &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

&lt;center&gt;Tab 3: Mapping quality&lt;/center&gt;

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.MAPQ.png" alt="Fig 8: MAPping Quality" width="35%" /&gt;
&lt;p class="caption"&gt;Fig 8: MAPping Quality&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## CIGAR (6)

.pull-left-60[

* Represents **how** the read aligned to the reference

* Use characters to represent individual operations


```default
# M Match
# N Gap
# D Deletion
# I Insertion
```

* An alignment with POS = `2` and  CIGAR = `6M`


```default
# REF: AAGTCTAGAA
# SEQ:   GTCTAG
```

* An alignment with POS = `2` and CIGAR = `3M2I3M`


```default
# REF: AAGT--TAGAA
# SEQ:  GTCGATAG
```

* An alignment with POS = `2` and CIGAR = `2M1D3M`


```default
# REF: AAGTCTAGAA
# SEQ:   GT-TAG
```

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.CIGAR.png" alt="Fig 9: CIGAR string" width="35%" /&gt;
&lt;p class="caption"&gt;Fig 9: CIGAR string&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## RNEXT (7), PNEXT (8), TLEN (9)

.pull-left-60[

* Reference sequence name of the primary alignment of the mate from a paired-end read

* Mapping position of the mate from a paired-end read

* Template length calculated from the mapping position from both mates

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.RNEXT.PNEXT.TLEN.png" alt="Fig 10: RNEXT, PNEXT, TLEN" width="45%" /&gt;
&lt;p class="caption"&gt;Fig 10: RNEXT, PNEXT, TLEN&lt;/p&gt;
&lt;/div&gt;

]

---

# The SAM fields

## SEQ (10), QUAL (11)

.pull-left-60[

* The sequence of the read

* The Phred quality scores of the read base calls

]

.pull-right-40[

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.SEQ.png" alt="Fig 11: Read sequence" width="85%" /&gt;
&lt;p class="caption"&gt;Fig 11: Read sequence&lt;/p&gt;
&lt;/div&gt;

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/sam/wgsim/reads.alignment.QUAL.png" alt="Fig 12: Phred quality score" width="85%" /&gt;
&lt;p class="caption"&gt;Fig 12: Phred quality score&lt;/p&gt;
&lt;/div&gt;

]

---

# The Samtools suite

## Overview

* Collection of programs for interacting with high-throughput sequencing data

* Contains three separate programs:

  1. Samtools
  
  2. BCFtools
  
  3. HTSlib

* Samtools is used to read and write SAM/BAM files

* The Samtools manual is available [online](https://www.htslib.org/doc/samtools.html)

* Developed by [Heng Li](https://www.broadinstitute.org/bios/heng-li) and [Bob Handsaker](https://www.broadinstitute.org/bios/bob-handsaker) et al. (2009)

* Alternative tools for SAM/BAM files include [BAMtools](https://github.com/pezmaster31/bamtools) and [Picard](https://github.com/broadinstitute/picard)

---

# The Samtools suite

.pull-left-50[

## Samtools commands

* Contains 36 different commands

* Grouped into different categories:

  * Indexing
  * Editing
  * File operations
  * Statistics
  * Viewing
  * Miscellaneous
  
* Convert SAM to compressed binary format called BAM

* Index SAM file to speed up retrieving alignments

]

.pull-right-50[

&lt;img src="data/sam/samtools-help.png" width="80%" style="display: block; margin: auto;" /&gt;

]

---

# Summary

- SAM stands for Sequence Alignment Map format

- Store biological sequences aligned to a reference sequence

- TAB-delimited text format consisting of a header and alignment section

- Each header has data fields for essential header information

- Each alignment has 11 mandatory fields for essential alignment information

- BAM is the compressed binary version of the SAM format

- Samtools is used to read, write, and execute SAM/BAM files

---

# Resources

## Specification

* [Sequence Alignment/Map Format](http://samtools.github.io/hts-specs/SAMv1.pdf)

## Manual

* [Samtools](https://www.htslib.org/doc/samtools.html)

## Papers

* [Twelve years of SAMtools and BCFtools](https://pubmed.ncbi.nlm.nih.gov/33590861) by Petr Danecek et al.

* [The Sequence Alignment/Map format and SAMtools](http://www.ncbi.nlm.nih.gov/pubmed/19505943) by Heng Li et al.

## Videos

* [Lockdown Learning: SAMtools](https://youtu.be/Llaxuzr6EkA) by Simon Cockell

* [Bioinformatics Coffee Hour: Samtools](https://youtu.be/Z4A2LCPyVU4) by Harvard FAS Informatics

* [SAM, BAM, CRAM format](https://youtu.be/XU8atPxM0VQ) by Aaron Quinlan

&lt;!-- --------------------- Do not edit this and below --------------------- --&gt;

---
name: end_slide
class: end-slide, middle
count: false

# Thank you. Questions?


.end-text[
&lt;p class="smaller"&gt;
&lt;span class="small" style="line-height: 1.2;"&gt;Graphics from &lt;/span&gt;&lt;img src="./assets/freepik.jpg" style="max-height:20px; vertical-align:middle;"&gt;&lt;br&gt;
Created: 20-Sep-2022 • James Ashmore • &lt;a href="https://www.zifornd.com"&gt;Zifo&lt;/a&gt;
&lt;/p&gt;
]


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
