<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Quality control</title>
    <meta charset="utf-8" />
    <meta name="author" content="James Ashmore • 20-Sep-2022" />
    <script src="libs/header-attrs-2.16/header-attrs.js"></script>
    <link href="libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <link href="libs/vembedr-0.1.5/css/vembedr.css" rel="stylesheet" />
    <link rel="stylesheet" href="assets/slide.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Quality control
]
.author[
### <b>James Ashmore</b> • 20-Sep-2022
]
.institute[
### Zifo RnD Solutions
]

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;
&lt;link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous"&gt;





&lt;!--------------- Only edit title, subtitle &amp; author above this ---------------&gt;


```r
knitr::opts_chunk$set(echo = FALSE, fig.align = "center")
```

---

## What is quality control?

* From Wikipedia, the free encyclopedia:

&gt; Quality control, commonly shortened to QC, refers to all those processes and procedures designed to ensure that the results of laboratory analysis are consistent, comparable, accurate and within specified limits of precision.

* In bioinformatics, quality control is performed at several stages:

    1. Quality of the raw data
    2. Effect of processing on the data
    3. Correctness of the code and software
    4. Interpretation of the results

* For sequencing data, we often start by looking at the quality of the raw data

* There are two types of quality metrics we can measure:

    * Generated by the sequencing platform (eg. quality scores)
    
    * Calculated directly from the raw reads (eg. base composition)

---

## Why is quality control important?

* In bioinformatics, quality control is necessary to avoid drawing the **wrong** conclusions

* There are many stages at which an error could be introduced:

    * The **code** you wrote is wrong
    * The **software** you used had a bug
    * The **data** you analysed was contaminated
    
* Errors in code and software are subtle and require due care and attention

* Problems with data are often less subtle and software can detect issues like contamination

* What happens if we **don't** take care in our analysis?

&lt;br&gt;

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/quality/xkcd_data_error.png" alt="xkcd: Data Error" width="65%" /&gt;
&lt;p class="caption"&gt;xkcd: Data Error&lt;/p&gt;
&lt;/div&gt;

---

## Retraction watch

*A tale of caution...*

.pull-left-50[

* The authors sequenced an ancient african genome from an individual they called "Mota"

* They compared "Mota" to other human genomes from across the globe (SNP analysis)

* Someone **forgot** to run a conversion script between one software and the next

* Data was inadvertently lost which made "Mota" appear less European than he actually was

* We don't penalize scientists for owning up to **mistakes**

* It is more important that we correct the scientific record!

]

.pull.right-50[

&lt;img src="data/quality/bioinformatics-error-1.jpg" width="32%" style="display: block; margin: auto;" /&gt;

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/quality/bioinformatics-error-2.webp" alt="Mota Cave" width="32%" /&gt;
&lt;p class="caption"&gt;Mota Cave&lt;/p&gt;
&lt;/div&gt;

]

---

## Sequencing reads

*A quick refresher on Illumina sequencing...*

<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/fCd6B5HRaZ8" width="889" height="500" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>

---

## Sequencing reads

*A quick refresher on FASTQ files...*

.pull-left-55[

* An extension of FASTA format to handle base quality metrics from sequencing machines

* Each sequence is represented by 4 lines:

  1. Identifier of the sequence starting with `@`
  2. Actual sequence
  3. Optional identifier starting with `+`
  4. Quality scores for each base
  
* FASTQ files are created and compressed with extension `.fastq.gz`

* For each sample per flow cell lane:

  * Single-end: `.fastq.gz`
  * Paired-end: `_1.fastq.gz` `_2.fastq.gz`
  
]

.pull-right-45[


```fastq
@HWUSI-EAS100R:6:73:941:1973#0/1
TGAAGNCTATAAACTAAGAAGCAAGCACACTAGGAGTT
+
AAAAA#EEEEA/EEEEEEE6EAEAEEEEEEEEEEEEEE
@HWUSI-EAS100R:6:73:942:1973#0/1
GTCACNATTCTCAAGGCCGTCGTCTTTTTAGTCGGTTT
+
AAAAA#EEEAEEAEEEEEEEAAAEEEEEE6E/E&lt;E/AE
@HWUSI-EAS100R:6:73:943:1973#0/1
GTCGCTAAGCTCTATATAGCGCGCGGGGGTATAGCTCG
+
AAAAA#EEEAEEAEEEEEEEEEEEEEEEE6E/E&lt;EEEE
```

]

---

## Quality control software

* There are multiple tools available to quality check sequencing reads

* In our *opinion* none of them are perfect, but there is a clear favorite which is [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

* Here's what Simon Andrews, the developer has to say:

&gt; FastQC aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.

* FastQC can be run as a GUI application or on the command line - *we recommend the latter!*

* You will learn to use FastQC in the workshop afterwards!

---

## FastQC

&lt;img src="data/quality/fastqc-preview.png" width="50%" style="display: block; margin: auto;" /&gt;

.pull-left-50[

**So how does FastQC work?**

* A series of analysis modules are run on a small sample of the sequencing data

* The module results are collected into a HTML report for inspection by the user

* Each module measures some aspect of the data which might reveal problems with the sequencing

]

.pull-right-50[

* A summary evaluation for each module is displayed:

    * Normal
    * Slightly abnormal
    * Very unusual
    
* Do not take these evaluations too seriously - *they are there as a rough guide only!*

]

---

## FastQC

### Basic Statistics

&lt;br&gt;

&lt;img src="data/quality/basic_statistics.png" width="90%" style="display: block; margin: auto;" /&gt;

&lt;p style="text-align: center;"&gt;&lt;i&gt;Sorry for the low-quality image!&lt;/i&gt;&lt;/p&gt;

---

## FastQC

### Per base sequence quality

&lt;img src="data/quality/per_base_quality.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Per tile sequence quality

&lt;img src="data/quality/per_tile_quality.png" width="80%" style="display: block; margin: auto;" /&gt;
---

## FastQC

### Per sequence quality scores

&lt;img src="data/quality/per_sequence_quality.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Per base sequence content

&lt;img src="data/quality/per_base_sequence_content.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Per sequence GC content

&lt;img src="data/quality/per_sequence_gc_content.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Per base N content

&lt;img src="data/quality/per_base_n_content.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Sequence length distribution

&lt;img src="data/quality/sequence_length_distribution.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Sequence duplicate levels

&lt;img src="data/quality/duplication_levels.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

### Overrepresented sequences

&lt;br&gt;

&lt;img src="data/quality/overrepresented_sequences.png" width="95%" style="display: block; margin: auto;" /&gt;

&lt;p style="text-align: center;"&gt;&lt;i&gt;Again, sorry for the low-quality image!&lt;/i&gt;&lt;/p&gt;

---

## FastQC

### Adapter content

&lt;img src="data/quality/adapter_content.png" width="80%" style="display: block; margin: auto;" /&gt;

---

## FastQC

*So what does 'good' data look like?*

&lt;iframe src="data/quality/fastqc-good.html" width="100%" height="500" data-external="1"&gt;&lt;/iframe&gt;

---

## FastQC

*And what does 'bad' data look like?*

&lt;iframe src="data/quality/fastqc-bad.html" width="100%" height="500" data-external="1"&gt;&lt;/iframe&gt;

---

## FastQC

*A demonstration by the one and only Simon Andrews...*

<div class="vembedr">
<div>
<iframe src="https://www.youtube.com/embed/bz93ReOv87Y" width="889" height="500" frameborder="0" allowfullscreen="" data-external="1"></iframe>
</div>
</div>

---

## Sequencing adapter trimming

.pull-left-50[

**Why do adapter sequences appear in sequencing reads?**

* Sequencing usually starts at the 5' end just after the primer

* The number of cycles dictates the number of bases sequenced

* Adapter read-through happens when the length of the DNA insert is less than the number of cycles

* No adapter read-through:

    * Insert length: 250
    * Sequencing cyles: 200
    * Adapter read-through: 200 - 250 = -50
    
* Adapter read-through:
    
    * Insert length: 100
    * Sequencing cycles: 200
    * Adapter read-through: 200 - 100 = 100


]

.pull-right-50[

&lt;br&gt;

&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="data/quality/adapter-origin.png" alt="Credit: ecSeq" width="90%" /&gt;
&lt;p class="caption"&gt;Credit: ecSeq&lt;/p&gt;
&lt;/div&gt;

]

---

## Sequencing adapter trimming

.pull-left-50[

**Why do we need to trim adapters from sequencing reads?**

* The presence of **artificial** sequences can interfere with read alignment:
    
    * Reads may be **misaligned** due to artificial sequence similarity
    
    * Reads may be **unaligned** due to reduced sequence similarity

* Trimming is essential for certain types of genomics data:

    * Genotyping
    
    * Genome assembly
    
    * Small RNA

* Trimming may not have a huge effect on other types of data, but why take the chance?

]

.pull-right-40[

&lt;br&gt;

&lt;img src="data/quality/why-adapter-trimming.jpeg" width="100%" style="display: block; margin: auto;" /&gt;

]

---

## Adapter trimming software

* There are multiple tools available to remove adapter sequences

* In our *opinion* there is not much difference between them

* They all have similar functionality and performance is comparable

* Some of the most popular tools include:

    * [cutadapt](https://cutadapt.readthedocs.io/en/stable/)
    
    * [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
    
    * [fastp](https://github.com/OpenGene/fastp)

* Good documentation is important, and for that reason we prefer **cutadapt**

* You will learn to use Cutadapt in the workshop afterwards!

---

## Adapter trimming software

* A quick look at the Cutadapt documentation

&lt;div class="figure" style="text-align: center"&gt;
&lt;iframe src="https://cutadapt.readthedocs.io/en/stable/" width="90%" height="475" data-external="1"&gt;&lt;/iframe&gt;
&lt;p class="caption"&gt;https://cutadapt.readthedocs.io/en/stable/&lt;/p&gt;
&lt;/div&gt;

---

## Adapter trimming software

* Cutadapt has a lot of convenient and intelligible features:

    * **Standard input and output**
    
        * Stream data without saving to disk
        * Save on storage and processing time

    * **Multi-core support**
    
        * Speed-up processing time of large files
        * Trimming less of a bottleneck in pipeline
    
    * **Multiple adapter types**
    
        * Remove 5' and 3' adapter sequences
        * Accommodates complex library preparation
    
    * **Quality trimming**
    
        * Remove low-quality base calls
        * Potentially improve alignment rate
    
    * **Read filtering**
    
        * Minimum read length
        * Discard both reads of a pair

---

## Adapter trimming software

* Cutadapt can detect multiple adapter types:

&lt;img src="data/quality/cutadapt-adapter-types.jpeg" width="85%" style="display: block; margin: auto;" /&gt;

---

## Adapter trimming software

* A quick look at the Illumina adapter sequences document:

&lt;div class="figure" style="text-align: center"&gt;
&lt;iframe src="https://support-docs.illumina.com/SHARE/AdapterSeq/illumina-adapter-sequences.pdf" width="90%" height="475" data-external="1"&gt;&lt;/iframe&gt;
&lt;p class="caption"&gt;https://support-docs.illumina.com/SHARE/adapter-sequences.htm&lt;/p&gt;
&lt;/div&gt;

---

## Adapter trimming software

* Cutadapt performs adaptive quality trimming:

&lt;img src="data/quality/cutadapt-trimming-algorithm.jpeg" width="80%" style="display: block; margin: auto;" /&gt;

---

## Adapter trimming software

* Trim a single 3’ adapter sequence:


```bash
cutadapt -a ADAPTER -o output.fastq input.fastq
```

* Trim multiple 3' adapter sequences:


```bash
cutadapt -a ADAPTER1 -a ADAPTER2 -o output.fastq input.fastq
```

* Trim low-quality bases from the 3' end:


```bash
cutadapt -q QUALITY -o output.fastq input.fastq
```

* Trim paired-end reads:


```bash
cutadapt -a ADAPTER1 -A ADAPTER2 -o out.1.fastq -p out.2.fastq in.1.fastq in.2.fastq
```

* Discard reads shorter than LENGTH:


```bash
cutadapt -m LENGTH -o output.fastq input.fastq
```

* Trim poly-A tails:


```bash
cutadapt -a "A{100}" -o output.fastq input.fastq
```

---

## Summary

* Quality control is a necessary step in all bioinformatics analysis

* It is there to prevent us drawing the "wrong" conclusions from the data

* For sequencing data, we start by looking at the quality of the reads

* FastQC is one such program you can use to quality check sequencing reads

* FastQC modules should be evaluated in the context of the experimental design

* Adapter contamination can be removed using read trimming software

* Cutadapt is one such program you can use, but they all perform equally well

* *Do not become paralysed by quality control! Know when to say good enough!*

&lt;!------------------------ Do not edit this and below -------------------------&gt;

---
name: end_slide
class: end-slide, middle
count: false

# Thank you. Questions?


.end-text[
&lt;p class="smaller"&gt;
&lt;span class="small" style="line-height: 1.2;"&gt;Graphics from &lt;/span&gt;&lt;img src="./assets/freepik.jpg" style="max-height:20px; vertical-align:middle;"&gt;&lt;br&gt;
Created: 20-Sep-2022 • James Ashmore • &lt;a href="https://www.zifornd.com"&gt;Zifo&lt;/a&gt;
&lt;/p&gt;
]



    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
